Configurable overlay position

koelle25 opened the pull request:
Hi Grayray75,

I really like the simplicity of your mod. But it was interfering with other mods (e.g. "BiomeInfo") because both render at the top left.

So I added an option to configure the position of the FPS display overlay. I built & tested it and it works like a charm. I also took care of handling the current GUI scale.

Actually I first tried it with a Slider 0-100% (using `@ConfigEntry.BoundedDiscrete(min = 0, max = 100)`). But as I couldn't find a way to determine current GUI resolution I just left it as a normal input field (to specify the pixel position). Maybe you know a way?

Anyway, I hope you'll like my addition.

Grayray75 commented on src/main/java/io/grayray75/fabric/fpsdisplay/mixin/InGameHudMixin.java:
```diff
@@ -22,8 +22,9 @@ public void render(MatrixStack matrixStack, float tickDelta, CallbackInfo info)
             String displayString = ((MinecraftClientMixin) client).getCurrentFPS() + " FPS";
-            float textPosX = 2;
-            float textPosY = 2;
+            int guiScale = client.options.guiScale;
+            float textPosX = config.offsetLeft / guiScale;
+            float textPosY = config.offsetTop / guiScale;
```
This will cause a crash if the GUI scale is set to 'auto', because then `client.options.guiScale` has the value 0

koelle25 replied:
Oh, I didn't test for that, thanks! In this case maybe it should use the values as is without scaling, what do you think?

I would change the code to this and add it as a commit to this PR:
```java
float textPosX = config.offsetLeft;
float textPosY = config.offsetTop;

int guiScale = client.options.guiScale;
if (guiScale > 0) {
  textPosX /= guiScale;
  textPostY /= guiScale;
}
```

Grayray75 commented:
Hi. After a bit of testing, I found a way to implement your first approach.
Can you take a look at it [here](https://github.com/Grayray75/FPS-Display/tree/feature/custom-position) and tell me if you like it?

koelle25 commented:
Hi, I just tested around with it. I noticed that 0-100 did not allow enough precision (especially horizontally) for me. This screenshot is taken at 1680x1050 (fullscreen) with GUI scale of 3 and the left offset is set to 1:
![Bildschirmfoto vom 2021-04-06 20-42-00](https://user-images.githubusercontent.com/915514/113762243-ac42cb80-9718-11eb-8e66-1050852a17ea.png)

The different offsets of the two texts were triggering my OCD :smile: So I removed the `BoundDiscrete` restriction and changed the `offsetTop` and `offsetLeft` variables to `float`.
Then I could set e.g. `0.5` as a value in the config screen and my perfectionism was satisfied:
![Bildschirmfoto vom 2021-04-06 20-47-18](https://user-images.githubusercontent.com/915514/113762842-4efb4a00-9719-11eb-910c-01f91e68b3d5.png)

If you want I can commit my current state into this PR.

Grayray75 commented:
Sorry for not replying for so long. I have been very busy recently.

> I just tested around with it. I noticed that 0-100 did not allow enough precision (especially horizontally) for me.
> So I removed the `BoundDiscrete` restriction and changed the `offsetTop` and `offsetLeft` variables to `float`.
> Then I could set e.g. `0.5` as a value in the config screen and my perfectionism was satisfied.
> 
I would like to leave the slider to adjust the values.
Do you think it would be precise enough if the sliders went from 0 to 200?

koelle25 commented:
Don't worry, that's okay. I know the feeling :grin: 

Yeah, I can see why you'd like the slider more - it's more user friendly. For me 0-200 would be enough precision.

The description texts in the lang files then will have to be adjusted to something more general, e.g.:
- Vertical Position (German: "Vertikale Position")
- Horizontal Position (German: "Horizontale Position")

Grayray75 commented:
Implemented in d372907
